### PR TITLE
[TG-2717] Generic inheritance with pointer replacement

### PR DESCRIPTION
**Pull request by majakusber**
When initiliazing a pointer and it is replaced by a different pointer (due to inheritance) we make sure that the generic parameter specialization map is updated too. For example, if we see a pointer to `Interface<Integer>` and there exists a class `A<T> implements Interface<T>` then we will be replacing the pointer with a pointer to `A<Integer>`. This PR makes sure that the generic type `A::T` is mapped to `Integer`.

Since CBMC does not construct replacement pointers, this change has no effect in CBMC itself (hence no unit tests or regression tests). The change is utilized in test generator: https://github.com/diffblue/test-gen/pull/1695.

### CODE REVIEW

File: src/util/std_types.h
Comment (tautschnig) [inline src/util/std_types.h]: Use `get_identifier()` instead of `get(ID_identifier)`. Looking at the code, `has_base` should actually also be amended in this way. With that looking-at-code done, I'm wondering whether this duplicate iteration over `bases` is a good idea? Can `get_base` actually be implemented by just doing the loop and `has_base` by doing `return bool(get_base(id));`?

File: src/java_bytecode/java_object_factory.cpp
Comment (thk123) [inline src/java_bytecode/java_object_factory.cpp]: Consider elaborating this comment with an example. E.g.:
```
// if replacement_pointer_type is Implementation<T> implements I<T, Float>
// then this inserts pairs:
//   I::T -> Implementation::T, 
//   I::U -> java.lang.Float
```

Assuming my understanding is correct?

File: src/java_bytecode/java_types.h
Comment (thk123) [inline src/java_bytecode/java_types.h]: Is there a reason to return an index (not yet looked at TG PR)? It seems like this code at least would be simpler to just return a `optionalt<reference_typet>`

File: src/java_bytecode/java_types.h
Comment (thomasspriggs) [inline src/java_bytecode/java_types.h]: NIT: I think this would be easier to read if the result of `std::find` was assigned to a temporary variable and checked against `generic_types().end()` in the `if` statement below, with the subtraction only done inside the `if` statement.

**Review by thomasspriggs — APPROVED** [review]
Just a minor readability comment, to fix if you have time.

File: src/java_bytecode/java_object_factory.cpp
Comment (majakusber) [reply src/java_bytecode/java_object_factory.cpp]: Not quite. Here is an example of what happens here: 
1. Assume we see a pointer to `I<Integer>`.
2. `convert_pointer_type` finds a class `Impl<T,U> implements I<T>` and builds a replacement pointer `Impl<Integer, U>` (note that CBMC does not build these replacement pointers, only test generator).
3. Below we simply input the pairs for the replacement pointer to the map (nothing inheritance-specific): `Impl::T -> Integer` (the second parameter is unspecialized).

File: src/java_bytecode/java_types.h
Comment (majakusber) [reply src/java_bytecode/java_types.h]: I will need the index later on in test generator so I thought a separate function for it would be useful. Would changing the name of the function make it more clear?

File: src/util/std_types.h
Comment (majakusber) [reply src/util/std_types.h]: Or rather `get_base(id).has_value()`?

Comment (tautschnig) [reply src/util/std_types.h]: Sure, that's likely the proper way of doing it!

**Comment by majakusber** [comment]
@thk123 Ready for re-review.

**Comment by thk123** [comment]
LGTM - just waiting on TG pointer approval

**Comment by majakusber** [comment]
@thk123 TG pointer passed CI, I rebased and squashed this PR.

**Review by thk123 — APPROVED** [review]
Looks like mostly transient problems, but if you could fix the clang format error then will merge once CI goes green

**Comment by majakusber** [comment]
Fixed, waiting for CI.

**Comment by majakusber** [comment]
@tautschnig Just noticed we need one more review for this from a code owner, would you mind taking a look at it again?

**Review by peterschrammel — APPROVED** [review]
LGTM

**Pull request merged**
